kernel/src/table_changes/schema_compat.rs

```diff
@@ -0,0 +1,302 @@
+use std::collections::{HashMap, HashSet};
+
+use crate::schema::{DataType, Schema, StructField, StructType};
+
+fn is_nullability_compatible(existing_nullable: bool, read_nullable: bool) -> bool {
+    // The case to avoid is when the read_schema is non-nullable and the existing one is nullable.
```

Comment:
Another possibility: Add these as methods on `StructType` itself? Then callers would be encouraged to do things like:
```rust
table_schema.can_read_as(read_schema)
```
... but I don't know a good way to do that for the nullability compat check since it's a plain boolean and doesn't always apply to a struct field (can also be array element or map value).

We could define a helper trait for struct/map/array, but that just pushes the problem to the trait impl (and there is only one call site for each type right now).